mypy/semanal.py

```diff
@@ -422,25 +422,82 @@ def is_defined_type_var(self, tvar: str, context: Context) -> bool:
+
+        # Decide whether to analyze this as a property or an overload.  If an
+        # overload, and we're outside a stub, find the impl and set it.  Remove
+        # the impl from the item list, it's special.
         t = []  # type: List[CallableType]
+        non_overload_indexes = []
         for i, item in enumerate(defn.items):
             # TODO support decorated overloaded functions properly
             item.is_overload = True
-            item.func.is_overload = True
             item.accept(self)
-            callable = function_type(item.func, self.builtin_type('builtins.function'))
-            assert isinstance(callable, CallableType)
-            t.append(callable)
-            if item.func.is_property and i == 0:
-                # This defines a property, probably with a setter and/or deleter.
-                self.analyze_property_with_multi_part_definition(defn)
-                break
-            if not [dec for dec in item.decorators
-                    if refers_to_fullname(dec, 'typing.overload')]:
-                self.fail("'overload' decorator expected", item)
 
-        defn.type = Overloaded(t)
-        defn.type.line = defn.line
+            if isinstance(item, Decorator):
+                item.func.is_overload = True
+                callable = function_type(item.func, self.builtin_type('builtins.function'))
+                assert isinstance(callable, CallableType)
+                if item.func.is_property and i == 0:
+                    # This defines a property, probably with a setter and/or
+                    # deleter.
+                    t.append(callable)
+                    self.analyze_property_with_multi_part_definition(defn)
+                    break
+                elif not any(refers_to_fullname(dec, 'typing.overload')
+                             for dec in item.decorators):
+                    if i == len(defn.items) - 1 and not self.is_stub_file:
+                        # Last item outside a stub is impl
+                        defn.impl = item
+                    else:
+                        # Oops it wasn't an overload after all. A clear error
+                        # will vary based on where in the list it is, record
+                        # that.
+                        non_overload_indexes.append(i)
+                else:
+                    t.append(callable)
+            elif isinstance(item, FuncDef):
+                if i == len(defn.items) - 1 and not self.is_stub_file:
+                    defn.impl = item
+                else:
+                    non_overload_indexes.append(i)
+        else:
+            # No break, so it was an overload.
+            if non_overload_indexes:
+                if t:
+                    # Some of them were overloads, but not all.
+                    for idx in non_overload_indexes:
+                        if self.is_stub_file:
+                            self.fail("Implementations of overloaded functions "
```

Comment:
Why the plural (here and in the next error)? Surely at most one implementation is allowed anyway. I think all three plurals in the sentence should become singulars.